coordinator/src/collaborative_revert.rs

```diff
@@ -28,254 +29,242 @@ use dlc_manager::Storage;
+/// transaction and would end up paying higher fees than necessary.
 const COLLABORATIVE_REVERT_TX_WEIGHT: usize = 672;
 
-pub async fn notify_user_to_collaboratively_revert(
-    revert_params: Json<CollaborativeRevert>,
-    channel_id_string: String,
-    channel_id: [u8; 32],
-    pool: Pool<ConnectionManager<PgConnection>>,
+/// Send a collaborative revert proposal to the counterparty in the channel identified by
+/// `channel_id`.
+///
+/// A collaborative revert involves signing a new transaction spending from the funding output
+/// directly. This can be used to circumvent bugs related to position and subchannel state.
+///
+/// The `ln_liquidity` argument is optional because one should rely on the values provided by the
+/// [`ChannelManager`] _if possible_. If LDK has already attempted to close the channel, it will be
+/// impossible to fetch its [`ChannelDetails`] and the caller will have to provide this data.
+#[allow(clippy::too_many_arguments)]
+pub async fn propose_collaborative_revert(
     node: Arc<Node<CoordinatorTenTenOneStorage, NodeStorage>>,
-    auth_users_notifier: mpsc::Sender<OrderbookMessage>,
-) -> anyhow::Result<()> {
-    let mut conn = pool.get().context("Could not acquire db lock")?;
+    pool: Pool<ConnectionManager<PgConnection>>,
+    sender: mpsc::Sender<OrderbookMessage>,
+    channel_id: [u8; 32],
+    settlement_price: Decimal,
+    fee_rate_sats_vb: u64,
+    funding_txo: OutPoint,
+    ln_liquidity: Option<LnLiquidity>,
+) -> Result<()> {
+    let mut conn = pool.get().context("Could not acquire DB lock")?;
 
-    let channel_details = node
-        .channel_manager
-        .get_channel_details(&channel_id)
-        .context("Could not get channel")?;
+    let channel_id_hex = channel_id.to_hex();
 
-    let sub_channels = node
+    let (ln_inbound_liquidity_sat, ln_outbound_liquidity_sat) =
+        match node.channel_manager.get_channel_details(&channel_id) {
+            Some(details) => {
+                let ln_inbound_liquidity_sat =
+                    Decimal::from(details.inbound_capacity_msat) / Decimal::ONE_THOUSAND;
+                let ln_inbound_liquidity_sat = ln_inbound_liquidity_sat
+                    .to_u64()
+                    .expect("inbound liquidity to fit into u64");
+
+                let ln_outbound_liquidity_sat =
+                    Decimal::from(details.outbound_capacity_msat) / Decimal::ONE_THOUSAND;
+                let ln_outbound_liquidity_sat = ln_outbound_liquidity_sat
+                    .to_u64()
+                    .expect("outbound liquidity to fit into u64");
+
+                (ln_inbound_liquidity_sat, ln_outbound_liquidity_sat)
+            }
+            None => {
+                let ln_liquidity = ln_liquidity
+                    .context("Unknown Lightning liquidity. Must provide that information")?;
+
+                (
+                    ln_liquidity.inbound_liquidity,
```

Comment:
❓ So basically if we do not have the channel details we have to provide the channel capacity split ourselves? That could be very failure prune, I guess we'd have to consider the reserve here as well?

Reply:
Yes, if the `ChannelDetails` cannot be fetched from the `ChannelManager`, we will have to provide the liquidity split through the API. A "simple" way to get this information might be to look for it in a backed up version of the coordinator.

There is no perfect solution here. This whole collaborative revert protocol is inherently imperfect.

Moving forward we could:

- Periodically save the liquidity split in the database.
- Modify LDK to print the `ChannelDetails` before erasing the channel.

> I guess we'd have to consider the reserve here as well?

I'm not sure what you mean, but the reserve should not be included in these values.

Reply:
I guess we could as well use the inbound and outbound liquidity stored on the shadow channel.